rust/src/storage/file/rename.rs

```
@@ -1,123 +1,27 @@
-                        "rename_noreplace failed with message '{}'",
-                        e
-                    )));
-                }
-                return Err(StorageError::other_std_io_err(format!(
-                    "failed to rename {} to {}: {}",
-                    from, to, e
-                )));
-            }
-            Ok(_) => Ok(()),
-        }
-    }
-
-    #[allow(unused_variables)]
-    unsafe fn platform_specific_rename(from: *const libc::c_char, to: *const libc::c_char) -> i32 {
-        cfg_if::cfg_if! {
-            if #[cfg(all(target_os = "linux", target_env = "gnu"))] {
-                cfg_if::cfg_if! {
-                    if #[cfg(glibc_renameat2)] {
-                        libc::renameat2(libc::AT_FDCWD, from, libc::AT_FDCWD, to, libc::RENAME_NOREPLACE)
-                    } else {
-                        // target has old glibc (< 2.28), we would need to invoke syscall manually
-                        unimplemented!()
-                    }
-                }
-            } else if #[cfg(target_os = "macos")] {
-                libc::renamex_np(from, to, libc::RENAME_EXCL)
-            } else {
-                unimplemented!()
-            }
-        }
-    }
-}
-
 /// Atomically renames `from` to `to`.
 /// `from` has to exist, but `to` is not, otherwise the operation will fail.
 #[inline]
 pub async fn rename_noreplace(from: &str, to: &str) -> Result<(), StorageError> {
-    imp::rename_noreplace(from, to).await
+    let from_path = String::from(from);
+    let to_path = String::from(to);
+
+    tokio::task::spawn_blocking(move || {
+        std::fs::hard_link(&from_path, &to_path).map_err(|err| {
+            if err.kind() == std::io::ErrorKind::AlreadyExists {
+                StorageError::AlreadyExists(to_path)
+            } else {
+                err.into()
+            }
+        })?;
+
+        std::fs::remove_file(from_path)?;
+
+        Ok(())
+    })
+    .await
+    .unwrap()
```

should we propagate the io error back to the caller instead?